src/coreComponents/linearAlgebra/utilities/AsyncRequest.hpp

```diff
@@ -0,0 +1,70 @@
+/*
+ * ------------------------------------------------------------------------------------------------------------
+ * SPDX-License-Identifier: LGPL-2.1-only
+ *
+ * Copyright (c) 2018-2020 Lawrence Livermore National Security LLC
+ * Copyright (c) 2018-2020 The Board of Trustees of the Leland Stanford Junior University
+ * Copyright (c) 2018-2020 TotalEnergies
+ * Copyright (c) 2019-     GEOSX Contributors
+ * All rights reserved
+ *
+ * See top level LICENSE, COPYRIGHT, CONTRIBUTORS, NOTICE, and ACKNOWLEDGEMENTS files for details.
+ * ------------------------------------------------------------------------------------------------------------
+ */
+
+/**
+ * @file AsyncRequest.hpp
+ */
+
+#ifndef GEOSX_LINEARALGEBRA_UTILITIES_ASYNCREQUEST_HPP_
+#define GEOSX_LINEARALGEBRA_UTILITIES_ASYNCREQUEST_HPP_
+
+#include "common/MpiWrapper.hpp"
+
+namespace geosx
+{
+
+/**
+ * @brief Helper class for managing non-blocking operation.
+ * @tparam T the type of the result of the non-blocking operation
+ */
+template< typename T >
+class AsyncRequest
```

Comment:
What about using https://en.cppreference.com/w/cpp/thread/async ?

Reply:
I was initially a bit skeptical about `std::async` because I was unsure about object lifetimes (we need persistent addresses and lifetimes for both send/recv buffers and the `MPI_Request` object until the very end), but upon some further thinking, since the objects have to be heap-allocated anyway, it makes no problem just moving them into the completion function as it is passed to `std::async`.

Check out new implementations of `iDot` and `iDotMultiple` in `VectorBase` and `BlockVectorView` in the latest commit. It also removes `AsyncRequest` which is no longer needed. Thanks for the suggestion!